src/mem/ruby/protocol/chi/CHI-cache-actions.sm

```diff
@@ -2600,11 +2600,18 @@ action(Send_CompData, desc="") {
   bool is_rd_nsd := tbe.reqType == CHIRequestType:ReadNotSharedDirty;
   bool is_rd_unique := tbe.reqType == CHIRequestType:ReadUnique;
 
-  // if the config allows (or not caching the data) and line has no sharers
-  bool snd_unique_on_rs := (fwd_unique_on_readshared || tbe.dataToBeInvalid)
+  // if the request type allows SD
+  bool snd_dirty_on_rs := is_rd_shared && !is_rd_nsd;
+
+  // Send UC/UD on ReadShared or ReadNotSharedDirty if the line has no sharers
+  // and one of the followings are met
+  // 1) the config allows or
+  // 2) local cache won't have the line or
+  // 3) dirty will be passed
+  bool snd_unique_on_rs := (fwd_unique_on_readshared ||
+                            tbe.dataToBeInvalid ||
+                            snd_dirty_on_rs)
```

Comment:
could we change this to `(tbe.dataDirty && !snd_dirty_on_rs)` ?

I think we still want to send CompData_SD when `fwd_unique_on_readshared==false`. Only if the request doesn't allow it (`snd_dirty_on_rs==false`) we would ignore fwd_unique_on_readshared and send CompData_UD.

Reply:
@tiagormk What about changing as below.
```
  // if the config allows (or not caching the data) and line has no sharers
  // Send UC/UD on ReadShared or ReadNotSharedDirty if the line has no sharers
  // and one of the followings are met
  // 1) the config allows or
  // 2) local cache won't have the line or
  // 3) data is dirty but the request doesn't allow SD
  bool snd_unique_on_rs := (fwd_unique_on_readshared ||
                            tbe.dataToBeInvalid ||
                            (tbe.dataDirty && is_rd_nsd))
                           && tbe.dataUnique && tbe.dir_sharers.isEmpty();

  // if the request type allows SD
  bool snd_dirty_on_rs := (is_rd_shared && !is_rd_nsd) ||
                          (is_rd_nsd && snd_unique_on_rs);
```

Baseline assumption is 'ditry' is passed wherever possible,
- `tbe.dataDirty && is_rd_nsd` intends that if the line is dirty and the request is ReadNotSharedDirty, Unique should be sent because dirty will be passed.
- Make it more straightforward for passing dirty condition; If the request is ReadShared, then pass dirty. If the request is ReadNotSharedDitry, send dirty only if Unique will be sent. This implies dirty is passed wherever possible.

Reply:
Thanks for the updates ! Made a few tweaks in your code below:
```
  // if the config allows (or not caching the data) and line has no sharers
  // Send UC/UD on ReadShared or ReadNotSharedDirty if the line has no sharers
  // and one of the followings are met
  // 1) the config allows or
  // 2) local cache won't have the line or
  // 3) Dirty will be passed but the request doesn't allow SD
  bool snd_unique_on_rs := (fwd_unique_on_readshared ||
                            tbe.dataToBeInvalid ||
                            (tbe.dataDirty && is_rd_nsd))
                           && tbe.dataUnique && tbe.dir_sharers.isEmpty();

  // if the request type allows SD
  bool snd_dirty_on_rs := (is_rd_shared && !is_rd_nsd) &&
                                            !tbe.dir_ownerExists;
```
- removed `!tbe.dir_ownerExists` from `snd_unique_on_rs`. The ` tbe.dir_sharers.isEmpty()`  check already implies `!tbe.dir_ownerExists`.
- removed `(is_rd_nsd && snd_unique_on_rs)` from `snd_dirty_on_rs`, since `snd_dirty_on_rs` is only actually checked if snd_unique_on_rs==false.

Reply:
@tiagormk , I've updated the codes accordingly. Please check the changes.